cpp/include/cudf/ast/expressions.hpp

```diff
@@ -588,12 +588,11 @@ class tree {
    * @returns a reference to the added expression
    */
   template <typename Expr, typename... Args>
-  Expr const& emplace(Args&&... args)
+  std::enable_if_t<std::is_base_of_v<expression, Expr>, Expr const&> emplace(Args&&... args)
```

Comment:
Was this just tightening up requirements for better error-handling? Did you run into a specific problem without this?

Reply:
Yes, it was just tightening up requirements for better error-handling.

Reply:
I personally find adding a `static_assert` check in this case can achieve the same goal while offering clearer error information.

Reply:
I did look into the compile error for the case where the requirements failed, it was reasonable.
I changed to `enable_if` in case our consumers might want to use TMP of any sort.